### PR TITLE
Table Front End

### PR DESCRIPTION
**Pull request by ric-evans**
Here it is, the front end for MoU Dashboard v1:

- Records (rows) can be added, modified, and deleted
- Columns (fields) can be hidden, some are hidden by default
- When adding a new row (data/record), if you have an institution or labor selected, this is auto filled in the new entry
- The table can be exported as an .xlsx file
- There are multiple buttons below the table:
   - The refresh button grabs data that is programmatically added on the backend, like the value for "US / Non-US".
         - this is needed do to a current limitation of Dash
   - The "Snapshot" buttons aren't currently operative
- Dropdowns are conditionally filtered per @paolodesiati's notes
- The numeric data (FTE) fields don't have logic implemented currently
    - We will need to decide where this logic lives (not on the front end)

Technical Details:
- You can completely ignore `data_source.py` this has logic that will make it's way into the REST interface(s), also dummy code
- ~~`icecube_setup.py` is copied from keycloak-rest-services (https://github.com/WIPACrepo/keycloak-rest-services/issues/3)~~

screencapture-localhost-8050-1599084698925
screencapture-localhost-8050-1599084744414
screencapture-localhost-8050-1599084845175
screencapture-localhost-8050-1599084865341

### CODE REVIEW

File: web_app/utils/data_source.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        key=lambda k: (
+            k[_WBS_L2],
+            k[_WBS_L3],
+            k[_US_NON_US],
+            k[INSTITUTION_LABEL],
+            k[LABOR_CAT_LABEL],
+            k[_NAMES],
+            k[_SOURCE_OF_FUNDS_US_ONLY],
+        ),
+    )
```
Comment (dsschult) [inline web_app/utils/data_source.py:89]: consider sorting in-place with `table.sort()`

File: web_app/utils/icecube_setup.py [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+        "is_US": False,
+        "region": _EUROPE,
+    },
+    "Yale": {
+        "cite": "Dept. of Physics, Yale University, New Haven, CT 06520, USA",
+        "abbreviation": "YALE",
+        "is_US": True,
+        "region": _US_CAN,
+    },
+}  # type: Dict[str, InstitutionMeta]
```
Comment (dsschult) [inline web_app/utils/icecube_setup.py:366]: We should figure out how to not duplicate this list across projects.

File: web_app/tabs/i3_mno.py [hunk 3]
```diff
@@ -175,252 +191,385 @@ def layout() -> html.Div:
...
-        elif not labor:
-            dff = DF[DF["Institution"] == institution]
-        else:
-            dff = DF[DF["Institution"] == institution]
-            dff = dff[dff["Labor Cat."] == labor]
-        return dff.to_dict("records")
+def toggle_pagination(n_clicks: int) -> Tuple[str, str, bool, int, str]:
+    """Toggle whether the table is paginated."""
+    if n_clicks % 2 == 0:
+        return "Show All Rows", "secondary", True, 15, "native"
```
Comment (dsschult) [inline web_app/tabs/i3_mno.py:557]: This is where the default number of rows comes from?

I think 15 seems small, but that's just me.

Comment (ric-evans) [reply web_app/tabs/i3_mno.py:557]: That's right. I'll make it configurable

File: web_app/utils/icecube_setup.py (hunk 2 above)
Comment (ric-evans) [reply web_app/utils/icecube_setup.py:366]: have you seen https://github.com/WIPACrepo/keycloak-rest-services/issues/3?

File: web_app/utils/data_source.py (hunk 1 above)
Comment (ric-evans) [reply web_app/utils/data_source.py:89]: good catch

File: web_app/utils/icecube_setup.py (hunk 2 above)
Comment (ric-evans) [reply web_app/utils/icecube_setup.py:366]: https://github.com/WIPACrepo/keycloak-rest-services/pull/4

**Review by olivas — APPROVED** [review]

File: web_app/tabs/i3_mno.py [hunk 4]
```diff
@@ -384,43 +384,28 @@ def update_name_email_icon(name: str, email: str) -> str:
...
-    labor: List[Dict[str, str]],
-    n_clicks: int,
-    rows: List[Dict[str, str]],
+    labor: str,
+    rows: str,
     columns: List[Dict[str, str]],
 ) -> List[Dict[str, str]]:
     """Filter dropdown menu & new rows callbacks."""
-    ctx = dash.callback_context
+    print("HAAAANNNNKKK")
```
Comment (olivas) [inline web_app/tabs/i3_mno.py]: Clean up logging.

**Review by olivas — CHANGES_REQUESTED** [review]

**Review by olivas — APPROVED** [review]

File: web_app/tabs/i3_mno.py [hunk 5]
```diff
@@ -1,84 +1,19 @@
...
-from ..utils.styles import (
-    CENTERED_100,
-    HIDDEN,
-    SHORT_HR,
-    WIDTH_22_5,
-    WIDTH_30,
-    WIDTH_45,
-)
+from ..utils import data
+from ..utils.styles import CENTERED_100, WIDTH_45
```
Comment (olivas) [inline web_app/tabs/i3_mno.py]: Are we OK with relative imports?  I generally am and I don't think we have clear and set standards, but some people are opposed.

**Review by olivas — APPROVED** [review]

**Review by olivas — APPROVED** [review]

**Review by olivas — APPROVED** [review]

**Review by olivas — APPROVED** [review]

**Review by olivas — APPROVED** [review]

**Review by olivas — APPROVED** [review]

**Review by olivas — APPROVED** [review]

File: web_app/tabs/i3_mno.py (hunk 5 above)
Comment (ric-evans) [reply web_app/tabs/i3_mno.py]: I am. My reasoning is that `web_app` is a module, so relative imports are to be expected.

**Review by olivas — APPROVED** [review]

**Pull request merged**